core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -699,15 +699,19 @@ public ActualProperties visitExchange(ExchangeNode node, List<ActualProperties>
                     if (inputProperty.isEffectivelySinglePartition() && node.getOrderingScheme().isEmpty()) {
                         verify(node.getInputs().size() == 1);
                         verify(node.getSources().size() == 1);
-                        PlanNode source = node.getSources().get(0);
-                        StreamPropertyDerivations.StreamProperties streamProperties = StreamPropertyDerivations.derivePropertiesRecursively(source, plannerContext, session, types, typeAnalyzer);
-                        if (streamProperties.isSingleStream()) {
+                        List<LocalProperty<Symbol>> inputLocalProperties = inputProperty.getLocalProperties();
+                        if (!inputLocalProperties.isEmpty()) {
                             Map<Symbol, Symbol> inputToOutput = exchangeInputToOutput(node, 0);
-                            // Single stream input's local sorting and grouping properties are preserved
-                            // In case of merging exchange, it's orderingScheme takes precedence
-                            localProperties.addAll(LocalProperties.translate(
-                                    inputProperty.getLocalProperties(),
-                                    symbol -> Optional.ofNullable(inputToOutput.get(symbol))));
+                            inputLocalProperties = LocalProperties.translate(inputLocalProperties, symbol -> Optional.ofNullable(inputToOutput.get(symbol)));
+                        }
+                        if (!inputLocalProperties.isEmpty()) {
```

Comment:
I think it deserves comment why for empty input local properties we don't need to validate if stream is single